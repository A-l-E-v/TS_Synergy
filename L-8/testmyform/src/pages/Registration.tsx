import { useEffect, useState } from 'react';
import SimpleReactValidator from 'simple-react-validator';
import Login from '../components/Login';
import Email from '../components/Email';
import Password from '../components/Password';

interface RegistrationForm {
  login: string;
  email: string;
  password: string;
}

const Registration = () => {
  const [form, setForm] = useState<RegistrationForm>({ login: '', email: '', password: '' });
  const [validator] = useState<SimpleReactValidator>(new SimpleReactValidator());
  const [validatorMessage, setValidatorMessage] = useState<string>(''); // Состояние для сообщения об ошибке

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setForm({ ...form, [e.target.name]: e.target.value });
  };

  const handleBlur = (e: React.FocusEvent<HTMLInputElement>): void => {
    // Устанавливаем сообщение об ошибке валидатора
    validator.showMessageFor(e.target.name);
    setValidatorMessage(validator.message(
      e.target.name,
      form[e.target.name as keyof RegistrationForm],
      'required')); // Получения сообщения
  };

  async function handleSubmit(event: React.FormEvent<HTMLFormElement>): Promise<void> {
    event.preventDefault();
    const response = await fetch('http://localhost:5000/api/auth/register', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(form),
    });

    if (!response.ok) {
      console.error('Ошибка при отправке данных:', response.statusText);
      return;
    }

    const data = await response.json();
    console.log(data);
  }

  useEffect(() => {
    if (validator.allValid()) {
      validator.hideMessages();
    }
  }, [form, validator]);

  return (
    <div className="form-container">
      <form onSubmit={handleSubmit} className="container-form">
        <h1>Регистрация</h1>
        <Login
          value={form.login}
          onChange={handleChange}
          onBlur={handleBlur}
          validator={validator}
          validatorMessage={validatorMessage} // Сообщение об ошибке
        />
        <Email
          value={form.email}
          onChange={handleChange}
          onBlur={handleBlur}
          validator={validator}
          validatorMessage={validatorMessage}
        />
        <Password
          value={form.password}
          onChange={handleChange}
          onBlur={handleBlur}
          validator={validator}
          // Используем validatorMessage
          validatorMessage={
            validator.message('password',
              form.password, 'required|min:8')}
        />
        <button
          type="submit"
          className="submit-button"
          disabled={!validator.allValid()}>Зарегистрироваться!</button>
      </form>
    </div>
  );
};

export default Registration;
