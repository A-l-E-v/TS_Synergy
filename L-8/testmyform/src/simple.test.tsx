import { render } from '@testing-library/react';
import Registration from './pages/Registration';

// Простой snapshot-тест страницы регистрации
describe('Registration Page', () => {
  it('should match the snapshot', () => {
    const { asFragment } = render(<Registration />);
    expect(asFragment()).toMatchSnapshot();
  });
});
