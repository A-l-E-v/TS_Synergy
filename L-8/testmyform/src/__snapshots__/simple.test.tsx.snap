// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registration Page should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="form-container"
  >
    <form
      class="container-form"
    >
      <h1>
        Регистрация
      </h1>
      <div>
        <label
          for="login"
        >
          Login
        </label>
        <input
          id="login"
          name="login"
          placeholder="Введите логин"
          type="text"
          value=""
        />
        <div
          class="srv-validation-message"
        />
      </div>
      <div>
        <label
          for="email"
        >
          Email
        </label>
        <input
          id="email"
          name="email"
          placeholder="Введите email"
          type="email"
          value=""
        />
        <div
          class="srv-validation-message"
        />
      </div>
      <div>
        <label
          for="password"
        >
          Пароль
        </label>
        <input
          id="password"
          name="password"
          placeholder="Введите пароль"
          type="password"
          value=""
        />
        <div
          class="srv-validation-message"
        />
      </div>
      <button
        class="submit-button"
        disabled=""
        type="submit"
      >
        Зарегистрироваться!
      </button>
    </form>
  </div>
</DocumentFragment>
`;
